terraso_backend/apps/core/permission_rules.py

```diff
@@ -4,3 +4,39 @@
 @rules.predicate
 def allowed_to_change_group(user, group_id):
     return user.is_group_manager(group_id)
+
+
+@rules.predicate
+def allowed_to_delete_membership(user, membership_id):
+    from apps.core.models import Membership
+
+    # Users are deleting their own membership
+    if user.memberships.filter(pk=membership_id).exists():
+        return True
+
+    membership = Membership.objects.get(pk=membership_id)
+
+    # Group Managers can delete any Membership in their Groups
+    if user.is_group_manager(membership.group.id):
+        return True
+
+    return False
+
+
+@rules.predicate
+def allowed_group_managers_count(user, membership_id):
+    from apps.core.models import Membership
+
+    membership = Membership.objects.get(pk=membership_id)
+    is_user_manager = user.is_group_manager(membership.group.id)
+    managers_count = membership.group.memberships.managers_only().count()
+    is_own_membership = user.memberships.filter(pk=membership_id).exists()
+
+    # User is the last manager and a Group cannot have no managers
+    if managers_count == 1 and is_user_manager and is_own_membership:
+        return False
+
+    return True
+
+
+rules.add_rule("allowed_group_managers_count", allowed_group_managers_count)
```

Comment:
Question: How does this work `rules.add_rule(`? it seems the only rule using this is `allowed_group_managers_count`why is that?

Reply:
Good question, @josebui. 

We have 3 predicates in this file, and we use them in two different ways. The first 2 predicates are associated directly to the model permissions (add, change, delete). They are referenced on `Meta` class on proper models. The third predicate, added in this PR, is added to the `rules` rule set. 

To evaluate a rule on rule set, we can call it as `rules.test_rule('rule_name', params)`.

Some examples that might help understand it:
```python
>>> @rules.predicate
>>> def is_book_author(user, book):
...     return book.author == user
>>>
>>> rules.add_rule('can_edit_book', is_book_author)
>>> rules.test_rule('can_edit_book', adrian, guidetodjango)
```

Using it with Django models:
```python
class Book(RulesModel):
    class Meta:
        rules_permissions = {
            "add": rules.is_staff,
            "read": rules.is_authenticated,
        }

# This is equivalent to:
rules.add_perm("app_label.add_book", rules.is_staff)
rules.add_perm("app_label.read_book", rules.is_authenticated)
```

Reply:
Great, thanks for the explanation Caio